src/main/java/net/shrimpworks/unreal/archive/mirror/MirrorClient.java

```diff
@@ -22,41 +23,61 @@
  * progress).
  */
 public class MirrorClient implements Consumer<MirrorClient.Downloader> {
-
-	private final Deque<Content> content;
+	private final int retryLimit = 4;
+	private Deque<Content> content;
+	private Deque<Content> retryQueue;
 	private final Path output;
 	private final int concurrency;
 	private final ExecutorService executor;
 
 	private final Progress progress;
 
 	private final long totalCount;
-	private final CountDownLatch counter;
+	private CountDownLatch counter;
 
 	private volatile Thread mirrorThread;
 
 	public MirrorClient(ContentManager content, Path output, int concurrency, Progress progress) {
-		this.content = new ArrayDeque<>(content.search(null, null, null, null));
+		this.content = new ConcurrentLinkedDeque<>(content.search(null, null, null, null));
+		this.retryQueue = new ConcurrentLinkedDeque<>();
 		this.output = output;
 		this.concurrency = concurrency;
 
 		this.progress = progress;
 
 		this.totalCount = content.size();
-		this.counter = new CountDownLatch(content.size());
 
 		this.executor = Executors.newFixedThreadPool(concurrency);
 	}
 
 	public boolean mirror() {
 		this.mirrorThread = Thread.currentThread();
 
+		// limit number of retry cycles
 		try {
-			// kick off the initial tasks, subsequent tasks will schedule as they complete
-			for (int i = 0; i < concurrency; i++) next();
+			for (int retryCount = 0; retryCount <= retryLimit && content.size() > 0; retryCount++) {
+				if (retryCount > 0) {
+					System.err.printf("%nA total of %d download(s) failed, retrying (%d/%d)...%n", content.size(), retryCount, retryLimit);
+				}
+
+				// initialize counter
+				this.counter = new CountDownLatch(content.size());
+
+				// kick off the initial tasks, subsequent tasks will schedule as they complete
+				for (int i = 0; i < concurrency; i++) next();
 
-			// wait for all downloads to complete
-			counter.await();
+				// wait for all downloads to complete
+				counter.await();
+
+				if (retryQueue.size() > 0) {
```

Comment:
Rather than doing a `.size()` check, I'd prefer to stick to the `!list.isEmpty()` style. Similarly for the `content.size()` check on L58.